libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -467,14 +483,35 @@ struct intrinsic_invoker_impl<Ret, std::tuple<array_ptr<T>, array_ptr<U>, size_t
 
    template<then_type Then>
    static Ret translate_one(wasm_interface &wasm, Inputs... rest, Translated... translated, I32 ptr_t, I32 ptr_u, I32 size) {
-      auto mem = getDefaultMemory(intrinsics_accessor::get_context(wasm).code.instance);
-      size_t length = size_t(size);
-      if(!mem || ptr_t+sizeof(T)*length >= IR::numBytesPerPage*Runtime::getMemoryNumPages(mem)
-              || ptr_u+sizeof(U)*length >= IR::numBytesPerPage*Runtime::getMemoryNumPages(mem))
-         Runtime::causeException(Exception::Cause::accessViolation);
-      T *base_t = (T*)(getMemoryBaseAddress(mem)+ptr_t);
-      U *base_u = (U*)(getMemoryBaseAddress(mem)+ptr_u);
-      return Then(wasm, array_ptr<T>{base_t}, array_ptr<U>{base_u}, length, rest..., translated...);
+      const auto length = size_t(size);
+      return Then(wasm, array_ptr<T>(wasm, ptr_t, length), array_ptr<U>(wasm, ptr_u, length), length, rest..., translated...);
+   };
+
+   template<then_type Then>
+   static const auto fn() {
+      return next_step::template fn<translate_one<Then>>();
+   }
+};
+
+/**
+ * Specialization for transcribing  a array_ptr type, followed by a size, and  a null_terminated_ptr type in the
+ * native method signature
+ * This type transcribes into 3 wasm parameters: a pointer, length, and a char pointer and checks the validity of
+ * that memory range before dispatching to the native method
+ *
+ * @tparam Ret - the return type of the native method
+ * @tparam Inputs - the remaining native parameters to transcribe
+ * @tparam Translated - the list of transcribed wasm parameters
+ */
+template<typename T, typename Ret, typename... Inputs, typename ...Translated>
+struct intrinsic_invoker_impl<Ret, std::tuple<array_ptr<T>, size_t, null_terminated_ptr, Inputs...>, std::tuple<Translated...>> {
+   using next_step = intrinsic_invoker_impl<Ret, std::tuple<Inputs...>, std::tuple<Translated..., I32, I32, I32>>;
+   using then_type = Ret(*)(wasm_interface &, array_ptr<T>, size_t, null_terminated_ptr, Inputs..., Translated...);
+
+   template<then_type Then>
+   static Ret translate_one(wasm_interface &wasm, Inputs... rest, Translated... translated, I32 ptr_t, I32 size, I32 ptr_u) {
+      const auto length = size_t(size);
+      return Then(wasm, array_ptr<T>(wasm, ptr_t, length), length, null_terminated_ptr(wasm, ptr_u), rest..., translated...);
    };
```

Comment:
@wanderingbort mentioned and I confirmed this is not needed. You can remove it. The variadic templates take care of it.

Reply:
So do you mean this whole specialization can go away?

Reply:
yes ;)